persist/fs/index_write.go

```diff
@@ -0,0 +1,298 @@
+}
+
+type writtenIndexSegmentFile struct {
+	segmentFileType IndexSegmentFileType
+	digest          uint32
+}
+
+// NewIndexWriter returns a new index writer with options.
+func NewIndexWriter(opts Options) (IndexFileSetWriter, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+	bufferSize := opts.WriterBufferSize()
+	return &indexWriter{
+		opts:             opts,
+		filePathPrefix:   opts.FilePathPrefix(),
+		newFileMode:      opts.NewFileMode(),
+		newDirectoryMode: opts.NewDirectoryMode(),
+		readerBuf:        bufio.NewReader(nil),
+		fdWithDigest:     digest.NewFdWithDigestWriter(bufferSize),
+	}, nil
+}
+
+func (w *indexWriter) Open(opts IndexWriterOpenOptions) error {
+	var (
+		nextSnapshotIndex int
+		err               error
+		namespace         = opts.Identifier.Namespace
+		blockStart        = opts.Identifier.BlockStart
+	)
+	w.err = nil
+	w.blockSize = opts.BlockSize
+	w.start = blockStart
+	w.snapshotTime = opts.Snapshot.SnapshotTime
+	w.segments = nil
+
+	switch opts.FileSetType {
+	case persist.FileSetSnapshotType:
+		w.namespaceDir = NamespaceIndexSnapshotDirPath(w.filePathPrefix, namespace)
+		// Can't do this outside of the switch statement because we need to make sure
+		// the directory exists before calling NextSnapshotFileIndex
+		if err := os.MkdirAll(w.namespaceDir, w.newDirectoryMode); err != nil {
+			return err
+		}
+
+		// This method is not thread-safe, so its the callers responsibilities that they never
+		// try and write two snapshot files for the same block start at the same time.
+		nextSnapshotIndex, err = NextIndexSnapshotFileIndex(w.filePathPrefix, namespace, blockStart)
+		if err != nil {
+			return err
+		}
+
+		w.fileType = persist.FileSetSnapshotType
+		w.checkpointFilePath = snapshotPathFromTimeAndIndex(w.namespaceDir, blockStart, checkpointFileSuffix, nextSnapshotIndex)
+		w.infoFilePath = snapshotPathFromTimeAndIndex(w.namespaceDir, blockStart, infoFileSuffix, nextSnapshotIndex)
+		w.digestFilePath = snapshotPathFromTimeAndIndex(w.namespaceDir, blockStart, digestFileSuffix, nextSnapshotIndex)
+	case persist.FileSetFlushType:
+		w.namespaceDir = NamespaceIndexDataDirPath(w.filePathPrefix, namespace)
+		if err := os.MkdirAll(w.namespaceDir, w.newDirectoryMode); err != nil {
+			return err
+		}
+
+		w.fileType = persist.FileSetFlushType
+		w.checkpointFilePath = filesetPathFromTime(w.namespaceDir, blockStart, checkpointFileSuffix)
+		w.infoFilePath = filesetPathFromTime(w.namespaceDir, blockStart, infoFileSuffix)
+		w.digestFilePath = filesetPathFromTime(w.namespaceDir, blockStart, digestFileSuffix)
+	}
+	return nil
+}
+
+func (w *indexWriter) WriteSegmentFileSet(segmentFileSet IndexSegmentFileSet) error {
+	// Always close the files
+	defer func() {
+		for _, file := range segmentFileSet.Files() {
+			file.Close()
+		}
+	}()
+
+	if w.err != nil {
+		return w.err
+	}
+
+	segType := string(segmentFileSet.SegmentType())
+	if segType == "" || !fileSubTypeRegex.MatchString(segType) {
```

Comment:
nit: could you declare as a helper on `SegmentType()`:

```
func (t SegmentType) String() (string, error) {
  s := string(t)
  if t == "" || !fileSubTypeRegex.MatchString(s) {
    return "" , fmt.Errorf("...")
  }
  return s, nil
}
```

Reply:
For sure, much cleaner.